test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+        "unsuccessfully checks payload of certificates with the right exception for each\
+        \ failure case"
+
+verifiesOkComm :: CommitmentOpening -> Bool
+verifiesOkComm CommitmentOpening{..} =
+    verifyCommitment coCommitment
+
+notVerifiesBadComm :: BadCommitment -> Bool
+notVerifiesBadComm (getBadComm -> badComm) =
+    not . verifyCommitment $ badComm
+
+verifiesOkCommSig :: SecretKey -> Commitment -> EpochIndex -> Bool
+verifiesOkCommSig sk comm epoch =
+    let commSig = (toPublic sk, comm, sign SignCommitment sk (epoch, comm))
+    in verifyCommitmentSignature epoch commSig
+
+notVerifiesBadCommSig :: BadSignedCommitment -> EpochIndex -> Bool
+notVerifiesBadCommSig (getBadSignedC -> badSignedComm) epoch =
+    not $ verifyCommitmentSignature epoch badSignedComm
+
+verifiesOkOpening :: CommitmentOpening -> Bool
+verifiesOkOpening CommitmentOpening{..} =
+    verifyOpening coCommitment coOpening
+
+notVerifiesBadOpening :: BadCommAndOpening -> Bool
+notVerifiesBadOpening (getBadCAndO -> badCommsAndOp) =
+    not . uncurry verifyOpening $ badCommsAndOp
+
+emptyPayload
+    :: Monoid container
+    => (container -> ExceptT e PureToss a)
+    -> MultiRichmenStake
+    -> GtGlobalState
+    -> Bool
+emptyPayload pureToss mrs gtgs =
+    isRight $ tossRunner mrs gtgs $ pureToss mempty
+
+newtype GoodCommsPayload = GoodComms
+    { getGoodComms :: (EpochIndex, GtGlobalState, CommitmentsMap, MultiRichmenStake)
+    } deriving (Show, Eq)
+
+instance Arbitrary GoodCommsPayload where
+    arbitrary = GoodComms <$> do
+        -- These fields won't be needed for anything, so they can be entirely arbitrary.
+        _gsOpenings <- arbitrary
+        _gsShares <- arbitrary
+
+        -- The epoch used in the tests is generated separately to make sure it exists.
+        (epoch, m) <- arbitrary :: Gen (EpochIndex, MultiRichmenStake)
+        richmen <- getValid <$> (arbitrary :: Gen (ValidRichmenStake GenesisMpcThd))
+        let richmenIds = HM.keys richmen
+            mrs = HM.insert epoch richmen m
+
+        -- This is the list of richmen which will be used to make keys for commitments
+        -- in the argument 'CommitmentsMap'.
+        richmenWithCerts <- sublistOf richmenIds
+        richmenWithComms <- sublistOf richmenWithCerts
+
+        stableCerts <- HM.fromList <$>
+            mapM (\r -> (,) <$> pure r <*> (arbitrary :: Gen VssCertificate))
+                 richmenWithCerts
+
+        -- Only the simplest case is considered w.r.t. the 'lastKnownEoS' field:
+        -- 'checkCommitmentsPayload' is called with the same epoch as the 'lastKnownEoS'
+        -- in 'gsVssCertificate'.
+        -- This is because rolling back slots is and should be tested elsewhere.
+        _gsVssCertificates <- VssCertData
+            <$> (pure . EpochOrSlot . Right . crucialSlot $ epoch)
+            <*> pure stableCerts
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
+            <*> arbitrary
```

Comment:
I think unsafe functions (e. g. unsafe constructors) should be avoided as much as possible. And if it's impossible, there should be comments explaining why this unsafe is safe.

> does it matter if some unrelated invariants are violated? 

Yes, I think it does for two reasons.
• Function may rely on the fact that invariants are not violated. It's an implementation detail whether it does or not. Even if you know it doesn't, you shouldn't rely on it if it's not written in the function's contract. If later someone changes implementation, they will have to also change these tests, even though they didn't break anything. But tests shouldn't start failing because of changes which don't break contract of the function.
• Currently we have `mkCommitmentsMapUnsafe` which is defined as `CommitmentsMap`, so it doesn't check anything. It assumes that the caller has already verified all input data. But later we may add a sanity check there and use `error` if it fails.

Apart from these two points, I would like to add that this way of constructing `VssCertData` looks strange to me anyway. What you want to create is `VssCertData` with given set of certificates and EoS. You can create an empty `VssCertData` (there is `empty`), use `setLastKnownEoS` to set EoS and then use `insert` to insert all certificates. This way you don't rely on the internal representation of `VssCertData` at all. You can't violate invariants of `VssCertData` and this code will compile if we change number of fields in `VssCertData` or change order of fields. Actually I don't like that `VssCertData`'s constructor is exported at all, but that's a separate question. And note that combining `empty`, `insert` and `setLastKnownEoS` isn't something hard to do.
For `CommitmentsMap` it's harder, because you need to know `SecretKey` to generate a `SignedCommitment` with given `StakeholderId` (which is needed to use `mkCommitmentsMap`). So you will need to write multiple generators. If it appears to be too much work, it can be ignored for now.